torch/csrc/jit/autodiff.cpp

```diff
@@ -0,0 +1,100 @@
+#include "torch/csrc/jit/autodiff.h"
+
+#include "torch/csrc/jit/symbolic_variable.h"
+#include "torch/csrc/utils/functional.h"
+
+namespace torch { namespace jit {
+
+using value_list = std::vector<Value*>;
+
+std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_values) {
+  const auto build_sym_grad = [node](const std::vector<SymbolicVariable>& grads) -> std::vector<SymbolicVariable> {
+    auto inputs = node->inputs();
+    switch(node->kind()) {
+      case kadd:
+        return {grads[0], grads[0]};
+      case ksub:
+        return {grads[0], -grads[0]};
+      case kmul:
+        return {grads[0] * inputs[1], grads[0] * inputs[0]};
+    }
+    throw std::runtime_error(std::string("don't support differentiation of `") +
+                            node->kind().toString() + "`");
+  };
+  auto sym_grads = build_sym_grad(fmap<SymbolicVariable>(grad_values));
+  return fmap(sym_grads, [](const SymbolicVariable &v) { return v.value(); });
+}
+
+void differentiate(std::shared_ptr<Graph>& graph) {
```

Comment:
We will need a version that takes  `, std::vector<bool> requires_grad)`, feel free to do that in a later PR, or we can add it when we run into it.

Reply:
I wasn't sure how exactly do we want to represent that information, so I omitted that. I'll add it in a later PR so we don't have to go over all this code again